uint256.go

```diff
@@ -198,9 +198,29 @@ func (z *Int) AddMod(x, y, m *Int) *Int {
 	if m.IsZero() {
 		return z.Clear()
 	}
-	if z == m { // z is an alias for m  // TODO: Understand why needed and add tests for all "division" methods.
+	if z == m { // z is an alias for m and will be overwritten by AddOverflow before m is read
 		m = m.Clone()
 	}
+
+	// Fast path for m >= 2^192, with x and y at most slightly bigger than m.
+	// This is always the case when x and y are already reduced modulo such m.
+
+	if (m[3] != 0) && (m[3] >= x[3]) && (m[3] >= y[3]) {
```

Comment:
This only handles happy case, but maybe if `m[3] < x[3]` maybe it would be better to perform `x mod m` and then also use the more efficient implementation. Same for `y mod m`.

Reply:
Happy case indeed, and also the normal case for modular arithmetic. :)

The code falls back to the slower division-based code for other cases, and will always return a fully-reduced result.

Reply:
Ok, I can see now we cannot efficiently use SubOverflow to reduce `x` here.